File: 프로그래머스/3/132266. 부대복귀/부대복귀.js
function solution(n, roads, sources, destination) {
    const map = new Map();
    
    for(let [from, to] of roads) {
        if(!map.get(from)) map.set(from, []);
        if(!map.get(to)) map.set(to, []);
        map.get(from).push(to);
        map.get(to).push(from);
    }
    
    const queue = [];
    const arr = new Array(n+1).fill(-1);
    
    queue.push(destination)
    arr[destination] = 0;
    
    while(queue.length > 0) {
        const current = queue.shift();
        for(let next of map.get(current)) {
            if(arr[next] === -1) {
                queue.push(next);
                arr[next] = arr[current]+1;
            } 
        }
    }
    
    return sources.map((v)=>arr[v]);
}

File: 프로그래머스/3/132266. 부대복귀/부대복귀.java
import java.util.*;

class Solution {
    public int[] solution(int n, int[][] roads, int[] sources, int destination) {
        int[] answer = new int[sources.length];
        int[] arr = new int[n+1];
        Map<Integer, Set<Integer>> map = new HashMap<>();
        
        for(int i = 0; i < arr.length; i++) arr[i] = -1;
        
        for(int i = 0; i < roads.length; i++) {
            int start = roads[i][0];
            int end = roads[i][1];
            
            if(!map.containsKey(start)) map.put(start, new HashSet<>());
            if(!map.containsKey(end)) map.put(end, new HashSet<>());
            map.get(start).add(end);
            map.get(end).add(start);
        }
        
        Deque<Integer> queue = new ArrayDeque<>();
        arr[destination] = 0;
        queue.offer(destination);
        
        while(queue.size() > 0) {
            int current = queue.pop();
            for(int next : map.get(current)) {
                if(arr[next] == -1) {
                    arr[next] = arr[current]+1;
                    queue.offer(next);
                }
            }
        }
        
        for(int i = 0; i < sources.length; i++) {
            answer[i] = arr[sources[i]];
        }
        
        return answer;
    }
}